app/controllers/concerns/application_shared.rb

```diff
@@ -33,42 +33,55 @@ def set_taxonomy
+            determined_taxonomy = nil
+          else
+            determined_taxonomy = available.where(:id => params["#{taxonomy}_id"]).first
 
-          # in case admin asked for taxonomy that does not exist or is not accessible, we reply with 404
-          if determined_taxonomy.nil?
-            not_found _("%{taxonomy} with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
-            return false
+            # in case user asked for taxonomy that does not exist or is not accessible, we reply with 404
+            if determined_taxonomy.nil?
+              not_found _("%{taxonomy} with id %{id} not found") % { :taxonomy => taxonomy.capitalize, :id => params["#{taxonomy}_id"] }
+              return false
+            end
           end
-        elsif request.session["#{taxonomy}_id"].present? # non-admin who didn't not specify explicit context
-          determined_taxonomy = find_session_taxonomy(taxonomy, user)
-        else # non-admin user who didn't specify explicit context and does not have anything in session
+        elsif session.has_key?("#{taxonomy}_id") # session with taxonomy explicitly set to id or nil (any contexxt)
+          if session["#{taxonomy}_id"].present?
+            determined_taxonomy = find_session_taxonomy(taxonomy, user) # specific id
+          else
+            determined_taxonomy = nil # explicitly set any context
+          end
+        else # user who didn't specify explicit context and does not have anything in session
           determined_taxonomy = find_default_taxonomy(user, taxonomy)
         end
       else # UI request
-        if request.session["#{taxonomy}_id"].present?
+        if session["#{taxonomy}_id"].present?
           determined_taxonomy = find_session_taxonomy(taxonomy, user)
         elsif !user.admin? && available.count == 1
           determined_taxonomy = available.first
         end
       end
 
-      store_taxonomy(taxonomy, determined_taxonomy) if determined_taxonomy.present?
+      set_current_taxonomy(taxonomy, determined_taxonomy)
+      store_taxonomy(taxonomy, determined_taxonomy) unless api_request?
     end
   end
 
+  # determined_taxonomy can be nil, which means any context
   def store_taxonomy(taxonomy, determined_taxonomy)
+    # session can't store nil, so we use empty string to represent any context
+    session["#{taxonomy}_id"] = determined_taxonomy.try(:id) || ''
+  end
+
+  def set_current_taxonomy(taxonomy, determined_taxonomy)
     taxonomy.classify.constantize.send 'current=', determined_taxonomy
-    request.session["#{taxonomy}_id"] = determined_taxonomy.id
   end
 
   def store_default_taxonomies(user)
     ['location', 'organization'].each do |taxonomy|
       default_taxonomy = find_default_taxonomy(user, taxonomy)
-      store_taxonomy(taxonomy, default_taxonomy) if default_taxonomy.present?
+      set_current_taxonomy(taxonomy, determined_taxonomy)
```

Comment:
@ares, seems `determined_taxonomy` is not defined

Reply:
works well when changed to `default_taxonomy`

Reply:
yeah, I pushed the fixing commit yesterday evening, I did amend so please reapply the last patch, I was hoping to fix it before you realize :-) I see there are still some test failures so I'll look into that today, it should be no big change